### PR TITLE
Add support for Ionicons (symbols) to <l:Icon> and <l:task> components

### PR DESCRIPTION
**Pull request by janfaracik**
As part of this implementation, I've renamed references to the icon pack Ionicons to Symbol(s) - for example, previously you'd call `getIonicon(...)`, now you'd call `getSymbol(...)`. 

The way this implementation works is by using the existing classes/components and only extending/tweaking them where strictly necessary. I've spent a fair amount of time playing with different implementations (such as new classes/components/rewriting chunks of Icon related classes) and found that this was the easiest/simplest approach.

To use a symbol in your sidebar action or management screen item, you just need to reference the symbol name and prefix it with "symbol-", e.g. if you want to use the settings icon you'd do:

carbon

I've also added a translation layer (thanks to @NotMyFault for the idea!) that will translate traditional Tango icons into Jenkins Symbols automatically, so plugin developers can maintain backwards compatibility whilst also taking advantage of newer, modern symbols. There are currently five icons supported by the translation layer although I plan to add more in a later PR so as not to balloon this one up more.

---

As an example of Symbols support I've replaced the Settings icons across Jenkins with a new flat version, and I've also updated the Jenkins About icon. With the translation service, several other icons have changed automatically too - the build icon, the delete icon and the plugins icon.

image

I've also spent some time as part of this to look at how to clean up the Icon jelly code, I've moved it to Functions.java so that it isn't duplicated across `task.jelly` and `manage.jelly`.

---

Once the `icon` and `task` components support Jenkins Symbols I'll open up follow up PRs to update icons across Jenkins - a WIP picture of the Manage Jenkins PR can be seen below:

### Proposed changelog entries

* Several icons across Jenkins have been updated - the build icon, the delete icon, the settings icon, the about icon and the plugin icon.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@timja 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by NotMyFault — COMMENTED** [review]
An interesting approach to get a rid of Tango (and other) icons and switch to a contemporary solution 🎉 

I'm wondering if it would be possible and actually realistic to work with overrides again like it's done for GIF -> PNG -> SVG icons provided by core.
E.g. imagine my plugin baseline is 2.289 and I reference to `icon-gear icon-md` in my jelly files to access my imaginary plugin configuration. If I run on the baseline, Jenkins displays me a PNG of the icon, on a version newer than 2.308 a SVG and on 2.anyFutureRelease the ionicon. That way I can retain backwards compatibility with the third last LTS release, but still display an ionicon over a Tango SVG on an appropriate version.

Though that is only an imaginary scenario that plopped into my mind when thinking about it.

On a different note, that doesn't work out nicely with non-core icons in the management or sidebar view

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (timja) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java:288]: New code should be compliant, just mark as private and add a getter
```suggestion
        private String iconXml;
```

File: core/src/main/resources/lib/layout/icon.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/icon.jelly:46]: guessing this is what you meant?
```suggestion
    <st:attribute name="title" deprecated="true">title, deprecated use tooltip instead</st:attribute>
```

File: core/src/main/resources/lib/layout/symbol.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/symbol.jelly:4]: Code search shows this is only used in core we can just remove it after updating all usages
https://cs.github.com/?scopeName=All+repos&scope=&q=org%3Ajenkinsci+l%3Aionicon

File: core/src/main/resources/hudson/PluginManager/available.jelly
Comment (timja) [inline core/src/main/resources/hudson/PluginManager/available.jelly:42]: shouldn't this be moved to l:icon? along with the others?

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:80]: Looking at the code I'm pretty sure this won't work for plugins wanting to add their own symbol.

Probably okay for this PR but should be addressed in a follow up

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2295]: In some code path `iconGuess` can be null, either needs a null check, or callers adjusted in jelly to prevent this

Noticed on the update center page when installing plugins: http://localhost:8080/updateCenter/

**Review by timja — COMMENTED** [review]
Apart from what was already mentioned about jcasc icon being broken I also noticed Lockable Resources icon broken on the dashboard and on the Configure Clouds page the `Manage Nodes` link still has the old icon:

image

**Comment by timja** [comment]
> Changelog: N/A - Nothing user facing

The gear icon is updated

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (timja) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java:287]: Remove the suppression now

**Comment by janfaracik** [comment]
> An interesting approach to get a rid of Tango (and other) icons and switch to a contemporary solution 🎉
> 
> I'm wondering if it would be possible and actually realistic to work with overrides again like it's done for GIF -> PNG -> SVG icons provided by core. E.g. imagine my plugin baseline is 2.289 and I reference to `icon-gear icon-md` in my jelly files to access my imaginary plugin configuration. If I run on the baseline, Jenkins displays me a PNG of the icon, on a version newer than 2.308 a SVG and on 2.anyFutureRelease the ionicon. That way I can retain backwards compatibility with the third last LTS release, but still display an ionicon over a Tango SVG on an appropriate version.
> 
> Though that is only an imaginary scenario that plopped into my mind when thinking about it.

That's a really great suggestion :) It wasn't too much work so I've added a small translation function which will existing Tango icons and translate them automatically to Jenkins Symbols, preserving styling etc.  There are currently five icons supported by the translation layer although I plan to add more in a later PR so as not to balloon this one up more.

> On a different note, that doesn't work out nicely with non-core icons in the management or sidebar view 

Hmm they're working on my end oddly enough

image

What's the address you're accessing Jenkins on? I'm using `http://localhost:8080/jenkins/` and all seems okay.

**Comment by NotMyFault** [comment]
> > An interesting approach to get a rid of Tango (and other) icons and switch to a contemporary solution 🎉
> > I'm wondering if it would be possible and actually realistic to work with overrides again like it's done for GIF -> PNG -> SVG icons provided by core. E.g. imagine my plugin baseline is 2.289 and I reference to `icon-gear icon-md` in my jelly files to access my imaginary plugin configuration. If I run on the baseline, Jenkins displays me a PNG of the icon, on a version newer than 2.308 a SVG and on 2.anyFutureRelease the ionicon. That way I can retain backwards compatibility with the third last LTS release, but still display an ionicon over a Tango SVG on an appropriate version.
> > Though that is only an imaginary scenario that plopped into my mind when thinking about it.
> 
> That's a really great suggestion :) It wasn't too much work so I've added a small translation function which will existing Tango icons and translate them automatically to Jenkins Symbols, preserving styling etc. There are currently five icons supported by the translation layer although I plan to add more in a later PR so as not to balloon this one up more.
> 

Thanks for the addition! Definitely something (up to date) plugins will benefit from without the need to remove older LTS support 👏🏻 


> > On a different note, that doesn't work out nicely with non-core icons in the management or sidebar view 
> 
> Hmm they're working on my end oddly enough

I'm currently on a clean installation build off `9fed655`, other icons are working fine, bundled ones like others, as the credentials icon e.g..

Taking this one for example, the icon path seems to be messed up... somehow
```html
<img src="/static/11fc0d88/static/11fc0d88/plugin/emailext-template/images/mail-mark-read.svg" class="icon">
```
Changing it to
```html
<img src="/static/11fc0d88/plugin/emailext-template/images/mail-mark-read.svg" class="icon">
```
makes it work as expected.

**Comment by timja** [comment]
> What's the address you're accessing Jenkins on? I'm using http://localhost:8080/jenkins/ and all seems okay.

I'm using the core-pr-tester on http://localhost:8080/manage and I'm hitting the issue too

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2294]: I don't think we need this in the public API?
```suggestion
    @Restricted(NoExternalUse.class)
    public static Icon tryGetIcon(String iconGuess) {
```

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2317]:
```suggestion
    @Restricted(NoExternalUse.class)
    public static String tryGetIconPath(String iconGuess, JellyContext context) {
```

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2343]: no need to do this twice
```suggestion
            String icon = IconSet.tryTranslateTangoIconToSymbol(iconMetadata.getClassSpec());
            if (icon != null) {
                return icon;
            }
```

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:543]: no need to be part of the API I think? ) internal use only
```suggestion
    @Restricted(NoExternalUse.class)
    public static String tryTranslateTangoIconToSymbol(String tangoIcon) {
```

**Review by timja — COMMENTED** [review]
this will need a green ATH and bom run before merging

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (daniel-beck) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: This is public API available to plugins…?

Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: not used anywhere though

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (janfaracik) [reply core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java]: Ah I think we need this as otherwise the field seemingly isn't exported/can't be read by breadcrumbs.js - unless I'm wrong and there's a way around that?

**Comment by NotMyFault** [comment]
A few things I noticed while giving it another test:

- The new ionicons on `/manage` lose their color if you click on them. Not sure if that's intended to change the colors on interaction, yet looks a bit weird imo.
- Breadcrumbs use no/the default color:

**Comment by janfaracik** [comment]
> A few things I noticed while giving it another test:
> 
> * The new ionicons on `/manage` lose their color if you click on them. Not sure if that's intended to change the colors on interaction, yet looks a bit weird imo.
> * Breadcrumbs use no/the default color:
>   

Awesome, thanks for giving it another try :) I've updated it so those issues should be resolved now.

I think the eventual idea would be to either have the icons follow text colour _or_ to have each 'Manage Jenkins' icon have its own unique colour, although I'm not too sure on how easy that'd be to implement/whether it'd be worth the inconsistency.

---

I've been working on the follow up PR to this one, which will update the rest of the Manage Jenkins icons, it looks like this at the moment and it fits in a bit better with the rest of the updated Jenkins' design.

image

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: How do you know?

But really, could we perhaps be more careful when introducing new API to ensure we are only adding what we can commit to support longer term?

**Comment by NotMyFault** [comment]
> Awesome, thanks for giving it another try :) I've updated it so those issues should be resolved now.

Yup, works like a charm.

> I think the eventual idea would be to either have the icons follow text colour or to have each 'Manage Jenkins' icon have its own unique colour, although I'm not too sure on how easy that'd be to implement/whether it'd be worth the inconsistency.

Having an unique color for each option _may_ be a unneeded and introduces much more noise:


> I've been working on the follow up PR to this one, which will update the rest of the Manage Jenkins icons, it looks like this at the moment and it fits in a bit better with the rest of the updated Jenkins' design.

Looking nice so far while fitting great with the ionicons 🎉

**Comment by daniel-beck** [comment]
> Having an unique color for each option may be a unneeded and introduces much more noise

Feedback on same-color weather icons has been fairly negative because it makes it difficult to distinguish them.

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: Searched via code search and it’s not an API that would make sense for others to call.

I’ve gone through earlier and asked for lots of methods to be excluded from API, will double check again though

**Comment by NotMyFault** [comment]
> > Having an unique color for each option may be a unneeded and introduces much more noise
> 
> Feedback on same-color weather icons has been fairly negative because it makes it difficult to distinguish them.

I'm not opposed to use colors to differentiate between options, the example I used was just to show how it _may_ can introduce too much noise, if every ionicon has its own color.
Using different, yet similar, colors or gradients as background color for the circles for each group on `/manage`, as proposed on Jan's screenshot, could work out well.

**Comment by timja** [comment]
conflicts

**Comment by timja** [comment]
ATH running: https://github.com/jenkinsci/acceptance-test-harness/pull/739
Need an incremental for bom

**Review by uhafner — CHANGES_REQUESTED** [review]
I think we need to adapt JavaDoc in `hudson.model.Action` as well. And the documentation of `lib/hudson/summary.jelly`. An example with the new icons would be helpful here.

(Additionally I think we should document the ideas in our developer documentation. If we don't do it now we will do it never. Maybe there should be a list of allowed symbol names as well?)

Some questions:
- How do we integrate new icons (see comment in first ion icons PR)? Plugin authors definitely will want this and if we do not provide a simple way now we will get mix of different solutions. Additionally, it would be nice if we automatically get updates on changes in https://github.com/ionic-team/ionicons via Dependabot rather than copying files manually around. If we do install them in multiple PRs we will get an upgrade problem for plugins that depend on fixed LTS releases. 
- The material icons and font awesome icons currently are loaded as a single SVG sprite. Will it be a performance problem, if we load individual files now for every icon?

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (uhafner) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:107]: Seems that this replacement does not work for non ionicon icons? I tried with `stroke:currentColor` in my SVG sidebar links but they still appear in black. What is required to get the correct color automatically? It would be helpful if this could work out-of-the box here as well:

Bildschirmfoto 2022-01-30 um 13 01 56

Comment (janfaracik) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:107]: Sadly there isn't any support for plugins yet. For adding Symbols to Jenkins, it should just work by adding them to the `symbols` folder - some SVGs may need a little tweaking (eg manually updating the `fill` attribute if it hasn't already been set) but for the most part they should just work automatically.

**Comment by janfaracik** [comment]
> I think we need to adapt JavaDoc in `hudson.model.Action` as well. And the documentation of `lib/hudson/summary.jelly`. An example with the new icons would be helpful here.
> 
> (Additionally I think we should document the ideas in our developer documentation. If we don't do it now we will do it never. Maybe there should be a list of allowed symbol names as well?)

I'll do that 👍 I need to add a lot of documentation still (especially once #5923 is merged). I'm not sure where best placed the docs should be - eg in the https://plugins.jenkins.io/ui-samples-plugin/ plugin or on the normal Jenkins docs page.

> Some questions:
> 
> * How do we integrate new icons (see comment in first ion icons PR)? Plugin authors definitely will want this and if we do not provide a simple way now we will get mix of different solutions. 

There's a few ways of doing this for plugin developers:
* Some of the icons plugin developers use will automatically be translated into Jenkins Symbols without any work needed 
* Plugin developers should be able to reference the Symbols and use them in their plugins the same way Jenkins does
* Adding their own custom symbols gets a bit trickier, this isn't currently supported although it's something that will **definitely** be added, just need to come up with a good implementation for this (this was discussed a little earlier in the thread + on the recent Jenkins UX sig)

> * Additionally, it would be nice if we automatically get updates on changes in https://github.com/ionic-team/ionicons via Dependabot rather than copying files manually around. If we do install them in multiple PRs we will get an upgrade problem for plugins that depend on fixed LTS releases.

That would be very cool 👍 

> * The material icons and font awesome icons currently are loaded as a single SVG sprite. Will it be a performance problem, if we load individual files now for every icon?

The icons once loaded are stored in a Map, so after the first load they'll be in memory rather than having to keep hitting storage.

**Comment by timja** [comment]
> I'll do that 👍 I need to add a lot of documentation still (especially once #5923 is merged). I'm not sure where best placed the docs should be - eg in the [plugins.jenkins.io/ui-samples-plugin](https://plugins.jenkins.io/ui-samples-plugin/) plugin or on the normal Jenkins docs page.

In here probably:
https://www.jenkins.io/doc/developer/views/

Maybe some more dynamic stuff in ui:samples

**Comment by uhafner** [comment]
> The icons once loaded are stored in a Map, so after the first load they'll be in memory rather than having to keep hitting storage.

Ah, they are inlined on the sever side? I wasn't aware of that. Then we don't have multiple calls for each image, that is fine. (On the other hand that means the HTML document contains *n* copies of the same SVG, isn't that problematic as well when the number of jobs is very high?)

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.js:26]: few reasons in these links:

https://stackoverflow.com/a/38786469

https://www.jotform.com/blog/why-you-should-embed-svg-inline-98570/#

the consensus generally is if you have a lot of them then use a sprite or img tag. If they are small and not a huge number then inline them

**Comment by timja** [comment]
conflicts

**Comment by janfaracik** [comment]
> > The icons once loaded are stored in a Map, so after the first load they'll be in memory rather than having to keep hitting storage.
> 
> Ah, they are inlined on the sever side? I wasn't aware of that. Then we don't have multiple calls for each image, that is fine. (On the other hand that means the HTML document contains _n_ copies of the same SVG, isn't that problematic as well when the number of jobs is very high?)

Yep, inlined server side. Performance shouldn't be an issue, I can't imagine there'll ever be a page with more than fifty of them at once - and even then it should be fine, just looking at the HTML source for this PR theres more inlined SVGs than we'll have on Jenkins.

**Comment by uhafner** [comment]
> > > The icons once loaded are stored in a Map, so after the first load they'll be in memory rather than having to keep hitting storage.
> > 
> > 
> > Ah, they are inlined on the sever side? I wasn't aware of that. Then we don't have multiple calls for each image, that is fine. (On the other hand that means the HTML document contains _n_ copies of the same SVG, isn't that problematic as well when the number of jobs is very high?)
> 
> Yep, inlined server side. Performance shouldn't be an issue, I can't imagine there'll ever be a page with more than fifty of them at once - and even then it should be fine, just looking at the HTML source for this PR theres more inlined SVGs than we'll have on Jenkins.

But a page with 1000 jobs has already 1000 play + status + weather icons which sum up to 1000 * 1000 bytes = 1Meg, or am I missing something?

**Comment by janfaracik** [comment]
> > > > The icons once loaded are stored in a Map, so after the first load they'll be in memory rather than having to keep hitting storage.
> > > 
> > > 
> > > Ah, they are inlined on the sever side? I wasn't aware of that. Then we don't have multiple calls for each image, that is fine. (On the other hand that means the HTML document contains _n_ copies of the same SVG, isn't that problematic as well when the number of jobs is very high?)
> > 
> > 
> > Yep, inlined server side. Performance shouldn't be an issue, I can't imagine there'll ever be a page with more than fifty of them at once - and even then it should be fine, just looking at the HTML source for this PR theres more inlined SVGs than we'll have on Jenkins.
> 
> But a page with 1000 jobs has already 1000 play + status + weather icons which sum up to 1000 * 1000 bytes = 1Meg, or am I missing something?

Ah of course, my brain totally slipped that that page doesn't paginate - right now the current implementation uses SVGs too just with the `<use>` tag, which essentially duplicates the node from an SVG file and places it on to the page, so we shouldn't have any worse performance than what we have now :)

**Comment by timja** [comment]
conflicts

**Comment by janfaracik** [comment]
> @janfaracik you able to take a look at:
> 
> > I think we need to adapt JavaDoc in hudson.model.Action as well. And the documentation of lib/hudson/summary.jelly. An example with the new icons would be helpful here.

Done 👍 Added a line mentioning Symbols for Action.java, replaced the doc for Summary.jelly + removed the icon loading logic (as its now all in Icon.jelly)

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
@uhafner you able to take another look?

**Review by uhafner — APPROVED** [review]
Thanks for adding the documentation!

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by zbynek** [comment]
"Pipeline Syntax" and "Changes" icons are missing in 2.335, I think that comes from this PR. The src of the missing images is `/icon-help icon-md` and `/icon-notepad` respectively.

**Comment by NotMyFault** [comment]
We're tracking the regressions in https://issues.jenkins.io/browse/JENKINS-67832 and https://issues.jenkins.io/browse/JENKINS-67828

**Comment by janfaracik** [comment]
Investigating this

**Comment by timja** [comment]
finishing testing of a fix 😄

**Comment by timja** [comment]
https://github.com/jenkinsci/jenkins/pull/6282

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:76]: This undocumented change fails PCT for `secure-requester-whitelist` thanks to its (luckily unnecessary) `icon-set` test dependency: https://github.com/jenkinsci/secure-requester-whitelist-plugin/pull/9

File: core/src/main/java/hudson/Functions.java
Comment (daniel-beck) [inline core/src/main/java/hudson/Functions.java:2302]: Caused [SECURITY-2777](https://www.jenkins.io/security/advisory/2022-06-22/#SECURITY-2777)

**Pull request merged**
